Add named frames to CollisionObjects (v2)

felixvd opened the pull request:
### Description

This PR supercedes #1060 and implements #1041. It includes all the changes requested in the last thread and now targets the correct branch. It requires [this PR](https://github.com/ros-planning/moveit_msgs/pull/50) to `moveit_msgs`.

In short, this PR adds subframes to CollisionObjects so that one can write "move the tip of object A to somewhere on object B". In the animation below, the cylinder tip is moved to each side of the box in this manner:

![simplescreenrecorder-2019-04-17_10 08 11](https://user-images.githubusercontent.com/4535737/56265284-1b861600-6124-11e9-8bfd-9b02d00700d2.gif)

I made a test package [here](https://github.com/felixvd/moveit_tutorials/tree/add-named-frames), which you can run with `rosrun moveit_tutorials subframes_tutorial`. You can reproduce the animation above by entering the numbers 2, 3, 4, 5.

**Small changelog:**  
I added a version of `getFrameTransform` to `robot_state.cpp` which combines the previous `getFrameTransform` and `knowsFrameTransform`, so that retrieval of a frame does not require two subsequent searches. I also added `getFrameInfo`, which returns the name of the robot link that a CollisionObject or named frame is attached to. This is used for changing the frames of constraints in `validateConstraintFrames`, because only links that are part of the link model can be used for motion planning.

I included the changes to the rest of the codebase where the new `getFrameTransform` would be used. I can take them out and start a separate PR for that, but they did not seem to cause issues.

**Last questions:**  
- [x] (Changed) Is the static variable I added [here](https://github.com/ros-planning/moveit/pull/1439/files#diff-15d4dc0e9e1eee143f7576803dfe8ea2R998) and [here](https://github.com/ros-planning/moveit/pull/1439/files#diff-8aac12a54166575f8df07630635ac4ffR1967) used correctly? I was stumped on how to return a reference correctly in a case like this. I suspect this is not thread-safe. Would appreciate a check from @rhaschke and/or whoever feels confident.
- [x] (fixed by using a reference) I made `getFrameTransform` forward to `getFrameInfo`, discarding the robot link returned by `getFrameInfo`. Should the functions be kept separate? I am not sure if makes a big difference performance-wise. The difference is one string copy.
- [x] Naming 1: In https://github.com/ros-planning/moveit_msgs/pull/50, @henningkayser proposed `subframe_ids` and `subframe_poses` instead of `frame_names` and `named_frame_poses`. Is "subframe" misleading? 
- [x] (Done) Naming 2: Another suggestion is to change the name of the `GetConstraintValidity` service. I am not opposed, it is only used internally anyway. `validateConstraintFrames` is the current version.

The Travis build will fail because moveit_msgs is affected. Before merging, I will remove the simple Rviz visualization of the named frames (the red dots in the animation). The visualization will have to be part of a PR addressing #1122 .

Would love to get this merged ASAP, because the GSoC project will affect this area of the code significantly. Thanks for reviewing!

@v4hn @davetcoleman 

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Document API changes relevant to the user in the moveit/MIGRATION.md notes
- [ ] Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Decide if this should be cherry-picked to other current ROS branches
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

jonbinney commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -985,55 +987,111 @@ bool RobotState::clearAttachedBody(const std::string& id)
 {
-  if (!id.empty() && id[0] == '/')
-    return getFrameTransform(id.substr(1));
+  static Eigen::Isometry3d transform;  // TODO(felixvd): Is this safe??
```
@felixvd this static variable is shared across all calls to this function, right? So if someone does:
```
const Eigen::Isometry3d& transform1 = robot_state.getFrameTransform("foo");
const Eigen::Isometry3d& transform2 = robot_state.getFrameTransform("bar");
```
Then After those two lines of code, transform1 will have the transform for "bar", right? (they both will?) This makes sense but seems icky.....

felixvd replied:
That's my understanding, so I can't imagine this is right. What is the right way to do this though? A `unique_ptr`? Even though the return type is a reference? I don't know enough C++ to get through this efficiently on my own, so I'm throwing this out there.

jonbinney replied:
I think this is a limation of returning a reference. It works OK if there's a semi-permanent object in the class that you can return a reference to. It doesn't work for computed-on-the-fly values, since there is no long term value in memory to return a reference to.

I can see two options:
1. keep an `std::map` of transforms for every frame within an object. Return references to those transforms. Optionally, you could still do lazy computation of the transform, keeping a "ditry" marker or something, and only computing them as needed.
2. Switch to return-by-value for these transforms

jonbinney replied:
I actually lean more towards option (2). Returning a reference seems like a bit of optimization overkill for such a small object.

gavanderhoorn replied:
There's a high chance that compilers used will use RVO here anyway, so just returning a normal object is probably OK.

felixvd replied:
The original version of these functions already returned references(`const Eigen::Isometry3d& getFrameTransform(const std::string frame_id)'), so this change is not for optimization. It is to return the results of `knowsTransform` and `getFrameTransform` in a single function. I could overload the original getFrameTransform to `const Eigen::Isometry3d& getFrameTransform(const std::string frame_id, const bool transform_successfully_found)` or something, but this pattern made more sense to me. I can't point to a guideline, but I assumed it is good practice, since you can write compact lines like [this](https://github.com/ros-planning/moveit/pull/1439/files#diff-e03d4d985fd8f5e9565368bddcac8d96R309).

If there's no way to implement this cleanly then I'll make the new functions return the references too, but maybe someone has an idea.

@jonbinney For the record, the references in the "main" `getFrameTransform`/`getFrameInfo` function are set to objects in the RobotState that outlive the function. It's only in the wrapper functions that an intermediate object is created so it can be returned.

felixvd replied:
Two more thoughts, after I posted [an SO question](https://stackoverflow.com/questions/55756238/return-a-reference-from-a-function-that-takes-it-as-a-parameter) about this:

1) I think the fact that the reference is non-const was unsafe to start with, because it allows values in the `RobotState` to be changed.

2) Apparently C++ 17 introduced [std::optional](https://en.cppreference.com/w/cpp/utility/optional) for cases like this. I will try that out.

felixvd replied:
Done in https://github.com/ros-planning/moveit/pull/1439/commits/4915f38946acbe526e4011aee721f9d273dae890, so I am resolving this.

felixvd replied:
Thanks for the input!

jonbinney commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -79,13 +80,13 @@ class World
   struct Object
   {
-    Object(const std::string& id) : id_(id)
+    Object(const std::string& object_id) : id_(object_id)
```
Is this name change just for clarity? Should the internal member variable be changed to match?

felixvd replied:
Yes, it was requested in the last thread. I don't think the name of the member variable in the object needs to change.

jonbinney commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -76,10 +76,10 @@ void World::addToObject(const std::string& id, const std::vector<shapes::ShapeCo
   int action = ADD_SHAPE;
 
-  ObjectPtr& obj = objects_[id];
+  ObjectPtr& obj = objects_[object_id];
```
I don't have a strong opnion on the `id` to  `object_id` change , but it creates a lot of noise in the diff that doesn't seem related to the purpose of the PR.

felixvd replied:
These types of changes are mostly accumulated comments from the last PR. I agree that it's become pretty big. All the more reason to get it over with, in my opinion.

rhaschke replied:
I suggest addressing these variable renamings in another PR.

jonbinney commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -131,14 +131,63 @@ void World::ensureUnique(ObjectPtr& obj)
+bool World::knowsTransform(const std::string& object_id) const
 {
-  auto it = objects_.find(id);
+  // Check object names first
```
It looks like this code does what it's meant to, but i'm torn. It seems a bit sketchy to put both "objects" and "child objects" in the same object namespace. Based on the description of the PR, I was expecting these child frames to only make sense in the context of their parent objects. For example "grasp_pose" or "handle_pose" on a teapot aren't the name of an object at the top level. It feels like it would be cleaner to not treat these child frames "objects" at all; so it seems like a top level "knowsTransform" should only check for object transforms (and therefor wouldn't be too useful compared to  `hasObject()`.

felixvd replied:
See the mainline comment

jonbinney commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -131,14 +131,63 @@ void World::ensureUnique(ObjectPtr& obj)
+  throw std::runtime_error("No transform found for object_id");
+}
+
+std::string World::getObjectOwningFrame(const std::string& frame_name) const
```
Same comment here - i'd prefer that named frames within an object only make sense in the context of an object. So it wouldn't make sense to call `getObjectOwningFrame("handle_frame")` because i might have objects `"teapot1"` and `"teapot2"1 in the scene which both have a `"handle_frame"`

felixvd replied:
See the mainline comment about namespacing

felixvd commented:
Thanks for the second look!

>     2. Mixing the child frames into the top level object namespace makes a few parts of the API slightly more convenient for the user, but I think it's too likely to cause problems. I think it would be a lot cleaner API (and less code) if the named frames only makes sense in the context of a specific object. Thoughts?

@henningkayser proposed prefixing the named frames/subframes with the object name (which already needs to be unique), so they're namespaced. The named frame `tip` of `screwdriver` would be `screwdriver_tip` or `screwdriver/tip` in the scene. What do you think about that?

I don't know what you mean by "the named frames only make sense in the context of a specific object". Do you mean the user should get the frame transformation from the object to the object part, apply it to the pose they want to move the part to, and then plan with the new pose applied to the end effector? I do think that's too cumbersome to be useful.

jonbinney commented:
@felixvd yes, i was proposing requiring the user to manually lookup the object and child-frame transforms separately and chain them together. I was imagining something like:
```
transform = world.getTransform("object_foo") * world.getObject("object_foo").getTransform("child_frame_bar")
```
Which didn't seem so bad; but I can see that it would be even more convenient to have a single function call in the API as you suggest. Prefixing with the object name is interesting, but it doesn't feel very machine-friendly (it's certainly user-friendly). How about having a function like:
```
transform = world.getChildFrameTransform(object_name, chlid_frame_name)
```
And then within the world we can do the lookups and multiply the two transforms, and raise a runtime error if one of them doesn't exist?

felixvd commented:
I can add the function, but it would make things a lot easier to just write "move top of cup under spout" instead of multiplying transforms. Normal users don't handle raw transforms in ROS either, and I don't think they should be expected to. It's something that notoriously goes wrong and that I don't miss doing at all.

I'll implement an attempt at namespacing this week though.

jonbinney commented:
Sounds good - having the function that handles the namespacing for the user sounds like a good compromise.

felixvd commented:
I changed the naming from "named_frames" to "subframes". Subframes are always prefixed by their parent object's name, so the subframe "tip" of an object "cylinder" can be used as "cylinder_tip" in the planning scene. The namespacing is done in `world.cpp` and `attached_body.h`, in the respective `getTransform` function for `collision::world::object`s and AttachedBody objects.

In the current commit there is a bug that seems to cause a crash when the subframe name is shorter than the object's name (?), but I should be able to fix that soon. You can see it occurring for "box_top" in my test package, but not for the longer subframe names. Shortening them causes the error to occur for them, too.

jonbinney commented:
Thanks @felixvd ! It is going to be a few days before i have time to do my next review of this, but I will get to it as soon as I can.

felixvd commented:
Just documenting that [this branch](https://github.com/felixvd/moveit/tree/debug-simplified-namespacing) contains essentially the same commit that seems to be causing me trouble, except this one is fine. Should be done before the maintainer meeting tomorrow.

rhaschke commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -105,13 +106,19 @@ class World
+     *  Use these to define points of interest on the object to plan with
+     *  (e.g. screwdriver_tip, kettle_spout, mug_base).
+     * */
+    std::map<std::string, Eigen::Isometry3d> subframe_poses_;
```
You need to use an aligned allocator for the Eigen transforms!
What about directly using [`moveit::core::Transforms`](https://github.com/ros-planning/moveit/blob/7fdc7f35f763c5e6784b5a72665188b2426fc009/moveit_core/transforms/include/moveit/transforms/transforms.h#L61) or at least [`moveit::core::FixedTransformsMap`](https://github.com/ros-planning/moveit/blob/7fdc7f35f763c5e6784b5a72665188b2426fc009/moveit_core/transforms/include/moveit/transforms/transforms.h#L54)?

felixvd replied:
I don't know what an aligned allocator is, but I can replace the `std::map` with `moveit::core::FixedTransformsMap` if that works?

rhaschke commented:
Dear Felix, could you please clean up your commit history a little bit?
I suggest separating the core functional changes from variable renamings. All fixups should be squashed into corresponding commits. As of now, the PR is large and not very modular in its commits.

rhaschke commented:
> Just documenting that this branch contains essentially the same commit that seems to be causing me trouble, except this one is fine.

@felixvd, what problem(s) are you referring to? Travis failures?

felixvd commented:
Thanks for the review!

> Dear Felix, could you please clean up your commit history a little bit?
> I suggest separating the core functional changes from variable renamings. All fixups should be squashed into corresponding commits. As of now, the PR is large and not very modular in its commits.

I usually do squash/reorder before pushing, but I am not sure what the right way to do it is when commits were already pushed. Separate everything into multiple new PRs and abandon this one? Force push the branch to a clean version of commits? I assume that would interfere with the PR and review interface here on Github?
Please advise on the preferred method.

> it feels strange to access those transforms in the same flat namespace as all other transforms. This makes it both, more costly (because you need to search all objects) and less transparent. I support Jon's proposal to use a function like
> `transform = world.getObjectSubFrame(object_name, chlid_frame_name)`

I worry that this is too complicated for the common user. I am hoping for this functionality to be readily accessible, and end users should not need to multiply transforms. The `world` object is already on a deeper level than the average user who uses only `geometry_msg` objects with the `move_group` or `planning_scene_interface`.

Just to confirm: Is it the search through all the subframes that you find problematic, or the access through the top-level namespace?

If it is the latter, I don't know that this design decision would be consistent. We already let users specify a collision object name as a target for planning, although that requires a search operation through the scene as well. If we do not value this convenience, why not ask users to do `transform = world.getFixedTransform(object_id)`? (That function does not exist at the moment, but it could)

If it is the former, I am not sure that performance would be affected significantly. The search through object subframes always happens last, so robot links, CollisionObjects and AttachedBody objects are retrieved as fast as before. It will also only make a difference in scenes with many objects and frames. It might be premature optimization to favor this rather than an easier interface.

This is the current API. I hope we can agree that this is very readable, even for beginners, and that readability is important:
```cpp
geometry_msgs::PoseStamped ps;
tf2::Quaternion quaternion;
moveit::planning_interface::MoveGroupInterface group("panda_arm");

ps.header.frame_id = "box_hole";
quaternion.setRPY(0, (180.0 / 180.0 * M_PI), 0);
ps.pose.orientation = tf2::toMsg(quaternion);
ps.pose.position.z = 0.01;
group.setEndEffectorLink("cylinder_tip");
group.setPoseTarget(ps);
group.go();
```
The only alternative I could imagine (to avoid the top-level namespacing) is something like this:  
*(ignore the slightly incorrect syntax)*
```cpp
moveit::planning_interface::PlanningSceneInterface planning_scene_interface;

//ps.header.frame_id = "box_hole"
quaternion.setRPY(0, (180.0 / 180.0 * M_PI), 0);
ps.pose.orientation = tf2::toMsg(quaternion);
ps.pose.position.z = 0.01;
ps = planning_scene_interface.transformPoseHeaderFromSubframe(ps, object="box", subframe="hole");
// Returns a transformed ps such that ps.header.frame_id = "box"
group.setEndEffectorLink("cylinder_tip");
group.setPoseTarget(ps);
group.go();
```
But I find this less elegant, and likely much harder to find for new users.

> @felixvd, what problem(s) are you referring to? Travis failures?

On 8f31daa, planning to the child frame `box_top` fails, but on e8c5172 it is fine. When I added the changes from 8f31daa in [this branch](https://github.com/felixvd/moveit/tree/debug-simplified-namespacing) while adding only comments, I could not reproduce the issue.

felixvd commented:
For the protocol, @rhaschke and I had a quick Skype call to discuss this PR and some ideas.

- Robert raised the idea of creating a global transform map in the planning scene, rather than keeping two separate entities (robot state and collision::world), which now get even more complex with this PR. I agree that there is some acrobatics involved to make it work (like the `validateConstraintFrames` service), and that the separation into robot_state and `collision::world` is strange and likely not very performant. I did not dare to make any more drastic changes, but I think this kind of change makes a lot of sense. We agreed to make that discussion a separate issue and try to get this PR merge-ready first.
- We are both in favor of the API I proposed, but with a slash as a prefix separator instead of an underscore, so the example from my last post would read:
```cpp
ps.header.frame_id = "box/hole"
group.setEndEffectorLink("cylinder/tip");
```
- The above actually already works, because the lookup ignores the separator (it just jumps one character to find the subframe name)
- I will condense the commit history of this PR and force push a few prettier commits by next week.

felixvd commented:
As a detail that is open to comments, I wondered in the OP if the string copy in `getFrameInfo` [here](https://github.com/ros-planning/moveit/pull/1439/files#diff-15d4dc0e9e1eee143f7576803dfe8ea2R1008) is wasteful, because `getFrameTransform` does not need the string. Robert proposed returning the pointer to the robot link instead of the name, so the overhead would only be an assignment and not a string copy. However, because the link model is const, a std::pair object would have to be created, and I am not sure how much overhead that is. The resulting function declaration would be:

`std::pair<const Eigen::Isometry3d&, const LinkModel*> RobotState::getFrameInfo(const std::string& frame_id, bool& frame_found) const;`

rhaschke commented:
> However, because the link model is const, a std::pair object would have to be created.

Can't you just "return" (via arguments) a const pointer to the LinkModel, like so:
`const Eigen::Isometry3d& getFrameInfo(const std::string &frame_id, LinkModel* const &link, bool &frame_found) const;`
Note the order of const in the type specification.
https://stackoverflow.com/questions/11514688/passing-const-pointer-by-reference

felixvd commented:
Sounds great, just what I was looking for. I'll do that if it works.

felixvd commented:
As it turns out, with the order `const LinkModel* &link` everything compiles and works fine, but with the order `LinkModel* const &link`, the compiler would give this error:

```shell
/home/catkin_ws/src/moveit/moveit_core/robot_state/src/robot_state.cpp:1019:44: error: assignment of read-only reference ‘robot_link’
     robot_link = robot_model_->getRootLink();
                                            ^
/home/catkin_ws/src/moveit/moveit_core/robot_state/src/robot_state.cpp:1019:43: error: invalid conversion from ‘const moveit::core::LinkModel*’ to ‘moveit::core::LinkModel*’ [-fpermissive]
     robot_link = robot_model_->getRootLink();
```

Second, I fixed the confusing issue I had earlier where only std::cout messages changed the program outcome. I accidentally left the return value undefined after catching an exception in `knowsSubframeTransform`.

Finally, as requested by @rhaschke, I cleaned up the commit history into hopefully readable chunks. I recommend looking at each of those separately instead of all the changed files together. There are more detailed explanations in the commit messages, as well. Thanks for having a read through it!

BryceStevenWilley commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,100 @@ bool World::hasObject(const std::string& object_id) const
+      try
+      {
+        if (object.second->subframe_poses_.find(object_id.substr(prefix_length)) != object.second->subframe_poses_.end())
+          return true;
```
If I'm reading this correctly, this isn't actually checking if the object name is the same as in the `object_id`, just if it's the same length, i.e. If `screw` has a subframe `tip`, and `stick` doesn't have any subframes, a query for `stick/tip` will incorrectly return true.

felixvd replied:
Haha, wow, what an awful mistake. I added explicit checks.

BryceStevenWilley commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,100 @@ bool World::hasObject(const std::string& object_id) const
+      int prefix_length = object.second->id_.length() + 1;
+      try 
+      {
+        if (object.second->subframe_poses_.find(object_id.substr(prefix_length)) != object.second->subframe_poses_.end())
```
See above comment.

BryceStevenWilley commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +111,63 @@ class AttachedBody
 
+  /** \brief Get the fixed transform to a subframe on this body (not a transform to a visual or collision shape).
+   * The frame_name needs to have the object's name prepended (e.g. "screwdriver/tip" returns true if the object's
+   * name is "screwdriver"). Returns an identity transform if frame is not found. 
```
Documentation is not correct here: it throws instead of returning an identity transform or returning true.

felixvd replied:
Fixed

BryceStevenWilley commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +111,63 @@ class AttachedBody
+  /** \brief Get all named transforms (not the ones associated to visual or collision shapes)
+   * These are also technically fixed transforms, but changing the original getFixedTransforms function would be bad.
+   */
+  const std::map<std::string, Eigen::Isometry3d>& getSubframeTransforms() const
```
Nit (feel free to ignore): IMO `getAllSubframeTransforms()` is better.

felixvd replied:
Wouldn't mind, but `getTransforms()` already existed

BryceStevenWilley commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -998,41 +1000,67 @@ const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_
   if (robot_model_->hasLinkModel(frame_id))
     return true;
+
+  for (auto body : attached_body_map_)  // Check if an AttachedBody has a child frame with name frame_id
```
Nit: should consistently refer to subframes, not child frames.

BryceStevenWilley commented on moveit_core/kinematic_constraints/include/moveit/kinematic_constraints/utils.h:
```diff
@@ -201,6 +201,20 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
+ * @param [in] c The message to be validated (modified directly).
+ */
+bool validateConstraintFrames(const robot_state::RobotState& state, moveit_msgs::Constraints& c,
+                              bool validate_position = true, bool validate_orientation = true);
```
Any use case where someone would be validating position or orientation and not the other?

felixvd replied:
Maybe in theory, but not really useful to think about imo. Those parameters are vestigial, I'll take them out.

BryceStevenWilley commented on moveit_core/robot_state/src/conversions.cpp:
```diff
@@ -284,12 +293,23 @@ static void _msgToAttachedBody(const Transforms* tf, const moveit_msgs::Attached
-            t0 = state.getFrameTransform(aco.object.header.frame_id);
+          t0 = state.getFrameTransform(aco.object.header.frame_id, frame_found);
+          if (frame_found)
+            ;
```
Prefer `if (not frame_found) { // put canTransform call here }`.

felixvd replied:
Added, I don't like the if/else cascade without brackets anyway.

BryceStevenWilley commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +111,63 @@ class AttachedBody
+      throw std::runtime_error("A subframe with name " + frame_name + " was requested from object " + id_ + " but it did not exist.");
+    }
+    catch (std::exception exc) {;}
+  }
```
Won't you have weird return error issues if you do catch a `std::exception`? Why include the try-catch in this case?

felixvd replied:
Because the attached body's id can be longer than the frame_name, so the substring would index out of range.

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1478,32 +1484,34 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
 
-      // we need to add some shapes; if the message is empty, maybe the object is already in the world
+      // STEP 1.1: Get poses from world or the message. Report error if all empty.
+      collision_detection::CollisionWorld::ObjectConstPtr obj_in_world = world_->getObject(object.object.id);
```
It is used in line 1597 to get the subframes from the object.

felixvd commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,100 @@ bool World::hasObject(const std::string& object_id) const
+    return true;
+  else  // Then objects' subframes
+  {
+    for (const auto& object : objects_)
```
Fixed where I found them, but there might be a few more trailing.

felixvd commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -131,14 +138,28 @@ class World
   /** \brief Check if a particular object exists in the collision world*/
   bool hasObject(const std::string& object_id) const;
 
+  /** \brief Check if a frame or object with name object_id exists in the collision world*/
```
Done

felixvd commented:
I updated each of the commits as suggested by Bryce, and started changing from `std::map` to `FixedTransformsMap` in the WIP commit, but it looks like it's not that easy. It seems that the `ensureUnique` function in `world.cpp` tries to copy the object, but `FixedTransformsMap` is set to `boost::noncopyable`.

Compiler error for the WIP commit:
```shell
Errors     << moveit_core:make /home/catkin_ws/logs/moveit_core/build.make.049.log
/home/catkin_ws/src/moveit/moveit_core/collision_detection/src/world.cpp: In member function ‘void collision_detection::World::ensureUnique(collision_detection::World::ObjectPtr&)’:
/home/catkin_ws/src/moveit/moveit_core/collision_detection/src/world.cpp:131:30: error: use of deleted function ‘collision_detection::World::Object::Object(const collision_detection::World::Object&)’
     obj.reset(new Object(*obj));
                              ^
In file included from /home/catkin_ws/src/moveit/moveit_core/collision_detection/src/world.cpp:37:0:
/home/catkin_ws/src/moveit/moveit_core/collision_detection/include/moveit/collision_detection/world.h:82:10: note: ‘collision_detection::World::Object::Object(const collision_detection::World::Object&)’ is implicitly deleted because the default definition would be ill-formed:
   struct Object
          ^~~~~~
/home/catkin_ws/src/moveit/moveit_core/collision_detection/include/moveit/collision_detection/world.h:82:10: error: use of deleted function ‘moveit::core::Transforms::Transforms(const moveit::core::Transforms&)’
In file included from /home/catkin_ws/src/moveit/moveit_core/collision_detection/include/moveit/collision_detection/world.h:41:0,
                 from /home/catkin_ws/src/moveit/moveit_core/collision_detection/src/world.cpp:37:
/home/catkin_ws/src/moveit/moveit_core/transforms/include/moveit/transforms/transforms.h:61:7: note: ‘moveit::core::Transforms::Transforms(const moveit::core::Transforms&)’ is implicitly deleted because the default definition would be ill-formed:
 class Transforms : private boost::noncopyable
       ^~~~~~~~~~
/home/catkin_ws/src/moveit/moveit_core/transforms/include/moveit/transforms/transforms.h:61:7: error: use of deleted function ‘boost::noncopyable_::noncopyable::noncopyable(const boost::noncopyable_::noncopyable&)’
In file included from /usr/include/boost/noncopyable.hpp:15:0,
                 from /home/catkin_ws/src/moveit/moveit_core/transforms/include/moveit/transforms/transforms.h:43,
                 from /home/catkin_ws/src/moveit/moveit_core/collision_detection/include/moveit/collision_detection/world.h:41,
                 from /home/catkin_ws/src/moveit/moveit_core/collision_detection/src/world.cpp:37:
/usr/include/boost/core/noncopyable.hpp:34:7: note: declared here
       noncopyable( const noncopyable& ) = delete;
       ^~~~~~~~~~~
```

I wouldn't know how to fix this (I assume `ensureUnique` is there for a reason), but I know that the subframe transforms are only used when changing collision objects in the scene and once in `validateConstraintFrames` when the planning request is constructed. Since they are not used in collision checking and other high-iteration procedures, there might not be a performance gain to be had here, but I don't know much about memory issues.

davetcoleman commented:
I skimmed the changes and all the comments just now, looks like enough maintainers have opinions on how this should be done. Thanks for this big feature effort @felixvd, I approve this feature!

BryceStevenWilley commented:
@felixvd I'm more confused as to why `Transforms` is marked `boost::noncopyable`. I think @rhaschke primarily wanted the Eigen aligned allocators, so the `FixedTransformsMap` typedef should work okay.

felixvd commented:
I see. I'm not at my dev machine today, but if you can, I'd appreciate if you could give a quick example of how to write it out. I tried Transforms yesterday because I wasn't sure how to use the typedef and it might save me some time stumbling in the dark.

jonbinney commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,91 @@ bool World::hasObject(const std::string& object_id) const
+    return it->second->shape_poses_[0];
+  else  // Find within subframes
+  {
+    for (const std::pair<std::string, ObjectPtr>& object : objects_)
```
I really prefer calls of the form `getTransform(object_id, subframe_id)` because they avoid all this for loop stuff. These calls in the world class are used by people in inner loops of algorithms on robots. In this implementation, some calls will look through every child frame of every object before returning, even if they have nothing to do with that object. I think that is way to high a price to pay to get the use-friendly one-string-with-slashes naming format. @BryceStevenWilley @rhaschke do you disagree?

BryceStevenWilley replied:
I don't disagree, but I think that style of search could be done with the current interface, as long as collision object names don't have slashes. Just search for a slash, if it's there, everything before is the object name and everything after is the subframe.  Not sure what restrictions there are on collision object names, but I feel like saying they shouldn't include slashes isn't too strict. 

I also had the impression that there wouldn't be an exorbitant amount of subframes in the whole scene, maybe 2 or 3 for some objects of interest. I could be wrong there though.

jonbinney replied:
(I realized after i said this that string comparison is linear time anyway) but still don't like the idea of creating a parsed naming scheme within the planning scheme that looks like tf but isn't.

felixvd replied:
I can understand the concern, but I can't evaluate the performance impact. If we ask for slashes not to be in the object name, it would only be one linear `string::find` operation for each call to `knowsTransform`. `getTransform` is not affected if no subframes are used. Not using slashes in names is so common that I don't think it would even be perceived as a restriction. I moved the check for a slash separator before the for-loop in world and robot_state for now, so this can move one way or another.

Second, I understand that the concern is about the nested lookups instead of through a dedicated additional function. @rhaschke proposed (in our call, not yet in an issue) to combine the robot state and collision world into one class with a TF-like hierarchical TransformMap for all fixed transforms. That way, each subframe would just become another entry in the map, so the search through them would not be as heavy. The current unified interface would be easy to update behind the scenes.

Last, I don't think that the interface will be confusing to users. In fact, "looks and works sort of like TF" might well make it easier to understand, because they already know TF. But that's just the UI end of things, naturally.

BryceStevenWilley replied:
I agree with what felixvd and rhaschke are proposing, that'd definitely get around the performance issue, and would go a long way to resolving https://github.com/ros-planning/moveit/issues/202.

While I do agree that looks like TF does make it easier to understand, I think what @jonbinney is saying is that users might *think* it's actually TF, mistakenly trying to publish subframes through TF instead of going through the move group interface? Am I correct in that interpretation?

jonbinney replied:
Yes, that they might think it is TF, or that since the frames have slashes like TF, that other TF-like behavior might be supported.   

My issue with this was purely a preference though, and I'm convinced just by hearing that others are comfortable with it. Lets move forward with this PR.

felixvd commented:
I pushed a new set of commits with the following changes:

- Changed `std::map` to `FixedTransformsMap`
- Moved the check for a slash separator in the frame name before the for-loop in `knowsFrameTransform`/`getFrameTransform` in `world` and `RobotState`, so that it is not entered needlessly. This seems like a good compromise for the moment.

If you want an easier way to follow the diff to the last version, please check out the commits on [this branch](https://github.com/felixvd/moveit/tree/WIP-add-named-frames-to-collision-objects). The [test package](https://github.com/felixvd/moveit_tutorials/tree/add-named-frames) is updated as well (`rosrun moveit_tutorials subframes_tutorial`).

rhaschke commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1478,32 +1484,34 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
     {
-      std::vector<shapes::ShapeConstPtr> shapes;
-      EigenSTL::vector_Isometry3d poses;
+      collision_detection::World::Object object_to_attach(object.object.id);
```
I don't see a benefit of introducing this World::Object as a container for shapes, poses, and subframes.
Keeping the old individual variables produces much less noise in this PR.

felixvd replied:
I added that to express both the intention and the provenance/target of each of the operations throughout the next 100+ lines of code. I found that I had to scroll up and down quite a lot to understand all the assignments of shapes and poses.

With the individual variables, the first time they are used for their purpose is [150 lines further down](https://github.com/ros-planning/moveit/compare/master...ubi-agni:new-add-named_frames-to-collision-objects#diff-8aac12a54166575f8df07630635ac4ffR1647) :(

I would argue for leaving it in for readability. Processing the message isn't a frequent operation that needs to be optimized for performance.

rhaschke replied:
I cannot follow this argumentation: The old variables were introduced in line 1484+1485 and first filled in lines 1498+1499. Your new variable is introduced in line 1490 and filled in 1506+1507 - pretty much the same distance. Of course, the first call to `attachBody` is far away, but that didn't change as well. In the end, you just introduced another variable...

rhaschke commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1478,32 +1484,34 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
-          for (Eigen::Isometry3d& pose : poses)
-            pose = i_t * pose;
+          // Transform shape poses to the link frame
+          const Eigen::Isometry3d& inv_transform = robot_state_->getGlobalLinkTransform(link_model).inverse();
```
Great improvement in variable names! Thanks.

rhaschke commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -1582,75 +1577,100 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
-                                   detach_posture);
-        ROS_DEBUG_NAMED(LOGNAME, "Added shapes to object '%s' attached to link '%s'", object.object.id.c_str(),
+        {
+          touch_links = std::set<std::string>(std::make_move_iterator(object.touch_links.begin()),
```
I'm not sure about the move operator here. `object` is const and hence shouldn't support move.
However, as the compiler accepts this, looks like move becomes standard copy in this case?

rhaschke replied:
It wasn't done like this before, but why not appending touch links as well?

felixvd replied:
1) Tbqh this might have been a clang_tidy addition that I just accepted. I'm not deep on memory access.

2) True, that would make sense for the append operation. Let's make it another PR.

rhaschke replied:
It's part of my fork already.

felixvd commented:
Thanks for the review already by the way!

rhaschke commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -161,6 +182,13 @@ class World
+   * If the object already exists, this call will assign the new shapes to the object
+   * at the specified pose. Otherwise, the object is created and the
+   * specified shape is added. This calls addToObjectInternal(). */
+  bool replaceShapesInObject(const std::string& object_id, const std::vector<shapes::ShapeConstPtr>& shapes,
```
This is an unrelated addition. Please file a separate PR for it.

rhaschke commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -175,6 +203,14 @@ class World
 
+  /** \brief Set subframes on an object. */
+  bool setSubframesOfObject(const std::string& object_id,
+                              const std::map<std::string, Eigen::Isometry3d>& subframe_poses);
```
A map with non-aligned Eigen objects should not be used at all. I suggest to keep the second version only.

felixvd replied:
Alright. Will remove tomorrow.

rhaschke replied:
Done in my fork.

rhaschke commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,95 @@ bool World::hasObject(const std::string& object_id) const
+              return true;
+          }
+        }
+        catch (const std::exception& e) {;}
```
What kind of exception do you want to catch here?

felixvd replied:
The object name can be longer than the subframe name ("bit/tip" vs "cylinder"), so `substr` indexes out of range.

rhaschke replied:
Ah. My solution, using `boost::starts_with` handles this gracefully as well.

rhaschke commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,95 @@ bool World::hasObject(const std::string& object_id) const
+          // Compare both object name and subframe name
+          int name_length = object.second->id_.length();
+          if (object.second->id_ == object_id.substr(0,name_length))
+          {
```
This code continues to search the (remaining) object list, even if the object name matched, but a corresponding subframe wasn't present.
Assuming that we don't have so many free objects in a scene, I accept this linear search through the object list for now.

felixvd replied:
True. I'll add a break.

rhaschke replied:
Fixed in my fork.

rhaschke commented on moveit_core/collision_detection/src/world.cpp:
```diff
@@ -136,6 +136,95 @@ bool World::hasObject(const std::string& object_id) const
+
+const Eigen::Isometry3d& World::getTransform(const std::string& object_id) const
+{
+  // This is almost identical to the function below, except for the return.
```
The code duplication is not very nice...

felixvd replied:
I can't remember why I added this comment instead of combining them. If I can't remember tomorrow I'll merge.

rhaschke replied:
Done too.

rhaschke commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +112,73 @@ class AttachedBody
+   * name is "screwdriver"). Returns an identity transform if frame is not found. 
+   * Use hasSubframeTransform to check if it exists.
+  */
+  const Eigen::Isometry3d& getSubframeTransform(const std::string& frame_name) const
```
Implementations of such "big" functions such go to .cpp instead of header.

rhaschke commented on moveit_core/robot_state/include/moveit/robot_state/attached_body.h:
```diff
@@ -110,6 +112,73 @@ class AttachedBody
+   * The frame_name needs to have the object's name prepended (e.g. "screwdriver/tip" returns true if the object's
+   * name is "screwdriver").
+  */
+  bool hasSubframeTransform(const std::string& frame_name) const
```
Same here.

rhaschke commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -998,41 +1000,78 @@ const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_
   }
+
+  // If frame_id contains a separating slash, look through subframes of the AttachedBody objects
+  if (frame_id.find("/")!=std::string::npos)
```
Here, the order of searching for attached bodies and subframes is reversed!

felixvd replied:
I put it in front because checking for only the separator in the name is a pretty light operation. I understood this approach to be the result of the discussion further up.

rhaschke commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -998,41 +1000,78 @@ const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_
   std::map<std::string, AttachedBody*>::const_iterator jt = attached_body_map_.find(frame_id);
   if (jt == attached_body_map_.end())
   {
     ROS_ERROR_NAMED(LOGNAME, "Transform from frame '%s' to frame '%s' is not known "
```
This looks broken.

felixvd replied:
Can you explain?

felixvd commented:
Thanks for the copious comments. A quick one from me: [here](https://github.com/ros-planning/moveit/commit/4792069b25eb49c56148715d99cb5d6f82e8c04b#diff-9582c92ebb2c026094ca2349707db3a4R147) you changed `getFrameTransform` to look for "object_name/" instead of "/" and then the object name. I think looking for the separator before entering the loop is more efficient.

rhaschke commented:
> ... you changed getFrameTransform() to look for "object_name/" instead of "/" and then the object name. I think looking for the separator before entering the loop is more efficient.

Not really. Searching for the slash is linear as well. And it needs to proceed to the end of the string. In contrast, checking that the frame_id starts with "object_name/" can return with false on the very first mismatching letter. 
However, my conceptual rationale was to not introduce a constraint for object names (to not contain slashes). If there is a slash in existing codes' object names, with my version this should continue to work, while it will fail with yours.

I'm done with all the review now. All suggested changes are available in [my fork](https://github.com/ubi-agni/moveit/tree/new-add-named_frames-to-collision-objects). Please throughly review all the `fixup` commits. I suggest to adopt my fork, force-push to this PR branch and continue with more fixes here.
@jonbinney, @BryceStevenWilley, please have another look when @felixvd has updated this PR with my changes.

One more fundamental remark: You essentially decided to `validateConstraintFrames()` on the client side by providing a _service_ in the `MoveGroupInterface`. However, I think both, server side (`move_group`) and client side can (and should!) perform this validation directly by calling `kinematic_constraints::validateConstraintFrames()` instead of performing a RPC. 
The move_group should do it just before handing over the PlanningRequest to the planner. 
And the `MoveGroupInterface::validateConstraintFrames()` utility function should do it using `getCurrentState()`. What was your rationale to provide a service for this basic functionality?

felixvd commented on moveit_core/planning_scene/src/planning_scene.cpp:
```diff
@@ -74,9 +74,9 @@ class SceneTransforms : public robot_state::Transforms
       return true;
     if (frame[0] == '/')
-      return knowsObject(frame.substr(1));
+      return knowsFrame(frame.substr(1));
```
Mmh, I think "frame" supersets "object frame" and "object subframe" better, but I don't have a strong opinion.

rhaschke replied:
In my opinion, frame supersets any kind of frame. However, here we are referring about frames related to objects - it's main frame or subframes.

felixvd commented:
Lovely, thanks for the detailed review! I'll push the merged changes tomorrow, hopefully. Some responses until then:

> Not really. Searching for the slash is linear as well. And it needs to proceed to the end of the string. In contrast, checking that the frame_id starts with "object_name/" can return with false on the very first mismatching letter.

True, but it has to check for every object_name, while the slash search is only done once. I'm not the right person to judge code performance, so I'm happy to defer on this, but @BryceStevenWilley and @jonbinney might want to weigh in because this was discussed further up.

> However, my conceptual rationale was to not introduce a constraint for object names (to not contain slashes). If there is a slash in existing codes' object names, with my version this should continue to work, while it will fail with yours.

That's more convenient for backwards compatibility, true. Object names with slashes will still risk bugs in the future. I'll add a migration note.

> One more fundamental remark: You essentially decided to validateConstraintFrames() on the client side by providing a service in the MoveGroupInterface. However, I think both, server side (move_group) and client side can (and should!) perform this validation directly by calling kinematic_constraints::validateConstraintFrames() instead of performing a RPC.
> The move_group should do it just before handing over the PlanningRequest to the planner.
And the MoveGroupInterface::validateConstraintFrames() utility function should do it using getCurrentState(). What was your rationale to provide a service for this basic functionality?

I tried to integrate it into the `move_group` because it seemed the most central, but `move_group.cpp` appeared to consist only of capabilities, so I made one. I also wasn't very familiar with the codebase when I did it. When I put it up for discussion in an earlier thread, @davetcoleman suggested a planning adapter.

Can you clarify what you mean by "the move_group should do it"? Do you mean the `move_group_interface` or a line in the `plan_service_capability` of the `move_group`?

What do you mean by "do it using `getCurrentState()`"? I assume you suggest to make `validateConstraintFrames` in `move_group_interface.cpp` look like this?

```cpp
bool validateConstraintFrames(moveit_msgs::Constraints& constraints)
  {
    robot_state::RobotState rs = current_state_monitor_->getCurrentState();

    ROS_DEBUG_NAMED(LOGNAME, "Transforming constraints to robot link (if necessary)");
    // Checks if the goal is defined either for existing link_names or the names of AttachedBody objects
    // that are attached to the robot. The latter are transformed to link_names so the planner can deal with them.
    bool valid = true;
    return kinematic_constraints::validateConstraintFrames(rs, constraints);
  }
```

rhaschke commented:
On the implementation `validateConstraintFrames`:
My suggestion is to have an implementation in both, the MoveGroupInterface and the move_group node.
1. The move_group should do the standard work, i.e. validating constraints before passing them to the actual planner. Having it as a planning adapter is a very good idea, but it requires to configure the adapter in your planning pipeline.
2. The MoveGroupInterface implementation should only be used in rare cases. Your suggested implementation looks fine.

felixvd commented:
1. OK. I'll put it in the `plan_service_capability` then.

felixvd commented:
Putting it in there has the awkward side effect of changing the request, unless I make a copy of it. See [here](https://github.com/felixvd/moveit/commit/43eec44d3591078aff005d04bfb8af25c2dffe61#diff-d159b6951d7ad53d77baef31f8393a1bR64).

felixvd commented:
Here we go, third revision this week. Lots of improvements in Robert's fork. Small changelog:

- I left the `object_to_attach` copy in the `ProcessAttachedCollisionObject` for now, because readability weighs more than the memory in my eyes. Would welcome more opinions.
- Removed the `validateConstraintFrames` service and inserted calls into the plan service capability and the move group interface
- Minor changes in comments

Other comments:

- I'm not sure what happened [here](https://github.com/ubi-agni/moveit/commit/8a549d6ca9612992ce681ab6b1e6366ab5adac64#diff-15d4dc0e9e1eee143f7576803dfe8ea2R895), but I'll accept.
- I like the combination of `knows*`and `get*` functions that is now continued [in the robot model](https://github.com/ubi-agni/moveit/commit/c923af655d44428eb7a514225587a0edaf819e47#diff-345a7e15d452dde605213102452017b0R1109).
- The append operation for touch_links makes sense, even if it's not technically related

One problem I didn't know how to solve:

`PlanningScene` returns a const reference for `getCurrentState`, but `CurrentStateMonitor` a RobotStatePtr. I changed `validateConstraintFrames` to take a pointer, but would like to preserve the version with a reference. I couldn't make the bit below work though because I'm no good with pointers.

```cpp
bool kinematic_constraints::validateConstraintFrames(const robot_state::RobotState& state, moveit_msgs::Constraints& constraints)
{
  robot_state::RobotStatePtr state_pointer(boost::make_shared(state));
  return validateConstraintFrames(state_pointer, constraints);
}
```

felixvd commented:
To clarify for @BryceStevenWilley and @jonbinney, these are the main points that need to be reviewed, I think:

- The last commit does not work because the `move_group_interface` does not know the attached objects. If no one knows why or how to fix it, I would leave the service call in version so we can move forward.
- Decide if the `object_to_attach` in the `ProcessAttachedCollisionObject` in the PlanningScene should be discarded or kept for readability

jonbinney commented:
@felixvd thanks for the summary - I'll take a look at those things tomorrow.

rhaschke commented:
@felixvd, I looked through your modifications to my modifications and I'm generally fine with them.
But still, I don't see the benefit of using `object_to_attach` in the `PlanningScene::processAttachedCollisionObjectMsg()`. In order to facilitate the review and the decision to include or drop this particular change, I separated this change into its own commit [here](https://github.com/ubi-agni/moveit/tree/subframes2). Please reset your PR branch accordingly. If thisis done, we can refer to this particular commit to be explicitly reviewed, e.g. by @jonbinney or @BryceStevenWilley. 

There are also some minor changes to the comments.
Note that my branch still uses the old service capability.

Regarding `validateConstraintFrames`: You cannot / should not validate the constraints in  `constructMotionPlanRequest()`. To validate, one needs to provide a RobotState to validate against.
However, this RobotState isn't necessarily the current robot state! If one uses the `plan_service` capability, it's perfectly valid to provide the initial robot state within the MotionPlanRequest. This use case isn't yet considered by you.

I think the cleanest approach to modify the constraints is indeed providing a planning adapter at the move_group side.

rhaschke commented:
Please rebase the PR branch to master as well. There was recently a fix (#1438) that might interfere with this PR.

felixvd commented:
> But still, I don't see the benefit of using `object_to_attach` in the `PlanningScene::processAttachedCollisionObjectMsg()`.

Only readability. 

> In order to facilitate the review and the decision to include or drop this particular change, I separated this change into its own commit [here](https://github.com/ubi-agni/moveit/tree/subframes2). Please reset your PR branch accordingly. If this is done, we can refer to this particular commit to be explicitly reviewed, e.g. by @jonbinney or @BryceStevenWilley.

👍 Done. Note that the same object_to_attach logic is in a later commit too, though.

> If one uses the `plan_service` capability, it's perfectly valid to provide the initial robot state within the MotionPlanRequest. This use case isn't yet considered by you.

It's not implemented, no. Originally because I was wrestling to get to a working version at all, now because I am fundamentally confused about which robot states and planning scenes and planning scene monitors know what.

> I think the cleanest approach to modify the constraints is indeed providing a planning adapter at the move_group side.

I'm not opposed, but I think it will have to be in a separate PR. I will be busy for a while from next week, so I won't be able to add a new feature. This is the fifth revision this month, anyway.

jonbinney commented:
@felixvd thanks a ton for all the effort you've been putting into this. The actual subframe part in the world code looks fine, but I have to admit I don't feel familiar enough with the kinematic constraint and some other parts of the codebase this touches to understand all the possible effects and corner cases. I'm not sure what to do in this case. I can "approve" the PR in the sense of approving the parts of it that I do feel confident in, if @rhaschke and @davetcoleman think that is useful.

I also think that for something that touches so many parts of the code, tests should be included in the PR.

rhaschke commented:
Thanks, @jonbinney for reviewing this PR and approving the representation part. What is your opinion about https://github.com/ros-planning/moveit/pull/1439#discussion_r280725127, i.e. the introduction of a dummy `World::Object` as a container for shape and transform variables? I don't like it, because it clutters the PR and I don't see a benefit over using the variables directly.
@felixvd argues, that it increases readability. But, I argue that the container variable is misleading because it suggests that this object will be added as an attached object, which is not going to happen...
I have factored out the related changes into a separate commit: e2117613273bf1c572106b06251b695d78453805 to facilitate the review. We could also leave this as a separate PR to decide on later.

Regarding your formal approval: I suggest that you don't formally approve this PR, because that might suggest to other maintainers that this PR is ready for merging, which it isn't yet.

@felixvd, I might find some time next weekend, when traveling to ICRA, to finish this PR and provide a suitable planning request adapter.

rhaschke commented:
Dear @felixvd,

I realized a PlanningAdapter for this feature, which is a really elegant solution now. Please verify my branch `subframes-planning-adapter` from https://github.com/ubi-agni/moveit.git and finally force-push it. Then, I think, we can merge this.

felixvd commented:
Thanks a lot! I will try to cake care of it today. If you're in Montreal already email me for a beer :)

felixvd commented:
1. Big fan of the rename from `validateConstraintFrames` to `resolveConstraintFrames`. Could have thought of that when I named the commit "resolve constraint frames". 

2. I added a description for the planning adapter in the xml and added a line, because the name might not be evident without context.

3. All that said, the tutorial stub I wrote does not seem to work anymore. I don't see the adapter running. Does it run for you? Does it have to be added somewhere?

4. The order of the planning adapters changed. I assume that doesn't affect anything?

felixvd commented:
As discussed at dinner, the adapter needs to be added to the `moveit_config` before this feature is active, which is inconvenient. I could not find the code that forces all default adapters to be loaded, or the discussion about it. I think @v4hn and @davetcoleman were involved.

That said, this works when the adapter is configured, and I am good with merging it. The last force push was just a comment fix. Thanks for everyone's help :)

rhaschke commented:
> I added a description for the planning adapter in the xml and added a line, because the name might not be evident without context.

The description in the xml is fine. But please undo the extensive comments.

> All that said, the tutorial stub I wrote does not seem to work anymore. I don't see the adapter running. Does it run for you? Does it have to be added somewhere?

You need to add the PlanningAdapter to the pipeline definition in your robot's MoveIt config.

> The order of the planning adapters changed. I assume that doesn't affect anything?

No. Just moved time parameterization to the end, which is more natural I think. Doesn't change functionality, because timing was added last anyway.

rhaschke commented:
@felixvd Could you, please have a look at the latest fixup and squash it when you are OK?

felixvd commented:
I made almost the same fix you proposed at the same time. I just force-pushed it. It still has the context explanation in utils.h because it doesn't seem excessive to me.

felixvd commented:
Related: We discussed that the RobotState did not know the attached objects when I tried resolving the constraint frames locally. [This](https://github.com/felixvd/moveit/tree/WIP2-add-subframes) is the branch where the error happens if you want to look into it. The robot state is taken from the `current_state_monitor` in the `move_group_interface`, but it does not seem to contain the attached bodies.

rhaschke commented:
As discussed a state or planning scene monitor _in your client_ is always lacking behind, because scene changes are published at a low rate only. I'm not even sure the state monitor can notice attached collision objects at all. I think, it only listens to joint state updates.

rhaschke commented:
@felixvd, I merged https://github.com/ros-planning/moveit_msgs/pull/50. Let's see whether Travis succeeds now.

felixvd commented:
Curious. I will try to look at it tomorrow night.

rhaschke commented:
@felixvd, I fixed various issues (most of them due to my changes). Please have a look at my branch `subframes-planning-adapter` in https://github.com/ubi-agni/moveit.git.

felixvd commented:
Looks good to me just reading the code, but can you explain the change in [this commit](https://github.com/ubi-agni/moveit/commit/c5f7ac95e1836198e0e13ca7e7b65f12f18ebced)? This seems to change behavior from the original version, and is also not in line with the behavior of `getTransform` in the same file.

rhaschke commented:
The original code considered object frames to be known only if there is a unique shape associated with it. You changed this semantic in the first place:
[SceneTransforms::knowsObject](https://github.com/ros-planning/moveit/blob/d0e3651f0960611b0d88c759dd18638c922ba753/moveit_core/planning_scene/src/planning_scene.cpp#L90-L94)
[PlanningScene::knowsFrameTransform](https://github.com/ros-planning/moveit/blob/d0e3651f0960611b0d88c759dd18638c922ba753/moveit_core/planning_scene/src/planning_scene.cpp#L1916)
[PlanningScene::getFrameTransform](https://github.com/ros-planning/moveit/blob/d0e3651f0960611b0d88c759dd18638c922ba753/moveit_core/planning_scene/src/planning_scene.cpp#L1886-L1897)

felixvd commented:
True. I was thinking of the getTransform logic which just gave a warning. My bad.

I force-pushed your branch to let Travis run, but [this line](https://github.com/ros-planning/moveit/pull/1439/files#diff-15d4dc0e9e1eee143f7576803dfe8ea2R1019) displays an error message when using subframes. I would keep using 'hasLinkModel'/'getLinkModel' because there are never that many robot links and creating the motion planning request is not a frequent operation, so this should not be costly.

Also do you want the fixup commits combined before merging?

rhaschke commented:
> this line displays an error message when using subframes.

You are right. I wasn't aware of the verboseness of this method anymore...
Please revert to the hasLink/getLink pair.

rhaschke commented:
@felixvd, can you file your tutorial examples as a PR, please?

rhaschke commented:
@felixvd, I just noticed that I already filed the tutorial PR: https://github.com/ros-planning/moveit_tutorials/pull/326
Please review there.

davetcoleman commented:
So glad this major new feature was merged in!!
